### DISCUSSION TITLE
Creating matrix blocks in a draft entry

### DISCUSSION THREAD

**Discussion by janhenckens** [craftcms/cms#11879] [Q&A] [answered]
For a translation plugin I would like to create a draft version of an entry, which includes al fields/relations/blocks that the original entry has.

I'm parsing fields and translation text where needed, and then passing an array of handles and values back to `setFieldValues()`.  

That works when I use `Craft::$app->getDrafts()->saveElementAsDraft()` (all matrix blocks are created) but the entire entry is turned into a draft at that point (which is exactly what the function is called so that's on me :))

Another way I tried is creating the draft by calling `Craft::$app->getDrafts()->createDraft()`, which handles that part correctly but which doesn't create the Matrix blocks (field values are all parsed the same). `createDraft()` is probably the way to go but should I then be creating and saving matrix blocks myself and passing those back?

Any ideas on what the best approach to this would be?

**Comment by brandonkelly** [comment, marked as answer]
To be clear, you want the new draft entry to be unaffiliated with the original entry, right?

**Reply by janhenckens** [reply]
For now I have the section set up to save elements in all sites (like we have in most of our projects), and I'd like the new draft to be a draft of the current entry in the specific site, so basically what the "Create a draft" button is doing.

**Reply by brandonkelly** [reply]
In that case you should just need to pass the entry into `Craft::$app->drafts->createDraft()`, which is what happens when “Create a draft” is pressed.

**Reply by janhenckens** [reply]
I was setting the values on the element before creating draft (of forget to add $draft->setCanonical($element). A closer look at the code in ElementsController and now it seems to work. Thanks for helping me out @brandonkelly! Plugin should be on the store soon :)

**Reply by brandonkelly** [reply]
Sweet!
